aggregator/src/aggregator.rs

```diff
@@ -2524,19 +2536,26 @@ impl VdafOps {
                         ));
                     }
 
-                    if let Some(collect_job_id) = tx
-                        .get_collect_job_id::<L, Q, A>(
-                            task.id(),
-                            &batch_identifier,
-                            &aggregation_param,
-                        )
-                        .await?
+                    // Check if this collect job already exists, ensuring that all parameters match.
+                    if let Some(collect_job) = tx.get_collect_job::<L, Q, A>(&collection_id).await?
                     {
-                        debug!(collect_request = ?req, "Serving existing collect job UUID");
-                        return Ok(collect_job_id);
+                        if collect_job.batch_identifier() == &batch_identifier
+                            && collect_job.aggregation_parameter() == aggregation_param.as_ref()
+                        {
+                            debug!(
+                                collection_id = %collection_id,
+                                collect_request = ?req,
+                                "Collect job already exists"
+                            );
+                            return Ok(());
+                        } else {
+                            return Err(datastore::Error::User(
+                                Error::UnrecognizedCollectJob(collection_id).into(),
```

Comment:
I'm not sure `UnrecognizedCollectJob` is appropriate here -- IIUC this error case correctly, the collect job exists, the caller is attempting to modify/assert a different state for the existing collect job. That is, this codepath identifies an idempotency problem.

We could introduce a new error type, which is reflected back to the caller as an HTTP 409 Conflict for now.

Reply:
Good catch. I wrote up https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/issues/406 to discuss this in DAP, but HTTP 409 sounds good for now.

Reply:
So, it's easy enough to define the new error type and return it from this point in the collection request handler. However, I noticed that we already had this problem when creating aggregation jobs and we don't handle it very gracefully: if we try to PUT the same aggregation job ID twice, it fails with a postgres error about violating a uniqueness constraint, and that materializes as an HTTP 500. I don't think that's hard to fix, but this PR is already quite large, so I think I'd like to deal with adding tests and bugfixes for idempotence of PUTting collection jobs and aggregation jobs in a subsequent PR.

Reply:
- [ ] add test proving that PUT collection job is idempotent
- [ ] add test proving that mutating a collection job with PUT fails with expected error
- [ ] add test proving that PUT aggregation job is idempotent
- [ ] add test proving that mutating an aggregation job with PUT fails with expected error
- [ ] add test proving that POST aggregation share is idempotent

...or file an issue to do that in a subsequent PR, depending on consensus we reach in this thread

Reply:
Filed #998 for the items above.